### ISSUE TITLE
afl-qemu-trace-32 and afl-qemu-trace-64

### ISSUE THREAD

**Issue by ele7enxxh** [AFLplusplus/AFLplusplus#1928]
Hi,

Sometimes, I need to perform fuzz testing on both 32-bit and 64 bit programs simultaneously. The current compilation and installation will overwrite the previous product afl-qemu-trace binary. Can you add a suffix to distinguish between the two? For example: afl-qemu-trace-32 and afl-qemu-trace-64.

Thanks.

**Comment by vanhauser-thc** [comment]
and how would afl-fuzz know which qemu binary to call? and what if someone wants that for arm64, mips etc.?
that just adds complexity on top that is not helpful I think.
the easy solution:
```
make
cd ..
cp -r AFLplusplus AFLplusplus32
cd AFLplusplus32/qemu_mode
export CPU_TARGET=i386
./build_qemu_support.sh
```
and then when fuzzing your simply set `AFL_PATH` to either /path/to/AFLplusplus or /path/to/AFLplusplus2

**Comment by p21nc3** [comment]
Hi @vanhauser-thc agreed with your point of considering other architectures. but having an binary selection method (dynamically) would work for this. lets say having a configuration setting for the users to configure the archi's they want to target. give it a thought ✌️

**Comment by ele7enxxh** [comment]
Hi, @vanhauser-thc, I think afl-fuzz can automatically select a specific afl-qemu-trace(eg.  afl-qemu-trace-arm, afl-qemu-trace-aarch64) by checking the file architecture of the target binary.

**Comment by vanhauser-thc** [comment]
Yeah that would be possible. I have no time for this though as I work on something for a presentation at the ccc congress.
You can send a PR if you think it is an important thing to have.

**Comment by p21nc3** [comment]
> Hi, @vanhauser-thc, I think afl-fuzz can automatically select a specific afl-qemu-trace(eg.  afl-qemu-trace-arm, afl-qemu-trace-aarch64) by checking the file architecture of the target binary.
> 
> 

Hi @ele7enxxh, are you working on this ?

**Comment by ele7enxxh** [comment]
> > Hi, @vanhauser-thc, I think afl-fuzz can automatically select a specific afl-qemu-trace(eg.  afl-qemu-trace-arm, afl-qemu-trace-aarch64) by checking the file architecture of the target binary.
> 
> Hi @ele7enxxh, are you working on this ?

No, Do you have free time to do it?

**Comment by p21nc3** [comment]
@ele7enxxh Sure, on it!

**Comment by AlLongley** [comment]
Hi team! I am also facing the situation described in OP where I am dynamically changing between architectures/fuzzing multiple arch binaries on a system simultaneously.

Just checking, was "AFL_QEMU_CUSTOM_BIN" a proposed solution to this? 

https://github.com/AFLplusplus/AFLplusplus/blob/db23931e7c1727ddac8691a6241c97b2203ec6fc/src/afl-common.c#L351


I've currently built multiple "afl-qemu-trace" architectures saved to different filenames, but I'm not sure if I'm doing something wrong here as this doesn't seem to have any effect:

`AFL_QEMU_CUSTOM_BIN=/usr/local/bin/afl-qemu-trace-i386`

**Comment by AlLongley** [comment]
Just as a potential band-aid patch, I've gotten great mileage out of compiling the two separate architectures as vanhauser-thc recommends, but then moving each to a separate folder

https://github.com/AFLplusplus/AFLplusplus/issues/1928#issuecomment-1847137442


So now I have two binaries:
```
/tmp/q64/afl-qemu-trace
/tmp/q32/afl-qemu-trace
```

I can call with

`PATH=/tmp/q64/:$PATH afl-fuzz -Q  .....`

and

`PATH=/tmp/q32/:$PATH afl-fuzz -Q  .....`

ensuring that `afl-qemu-trace` is not in my $PATH such as /usr/bin/ etc.

Likely `AFL_PATH` is the proper AFL specific environment variable for this, but the above works fine for me
